File: bindings/csharp/RenderPipeline.cs
using System;
using System.Runtime.InteropServices;

namespace Astral.Canvas
{
    [StructLayout(LayoutKind.Explicit)]
    public struct BlendState
    {
        [FieldOffset(0)]
        Blend srcColor;
        [FieldOffset(4)]
        Blend destColor;
        [FieldOffset(8)]
        Blend srcAlpha;
        [FieldOffset(12)]
        Blend destAlpha;

        public BlendState(Blend src, Blend dest)
        {
            srcColor = src;
            destColor = dest;
            srcAlpha = src;
            destAlpha = dest;
        }
        /*Additive = new BlendState(Blend.SourceAlpha, Blend.One);
        AlphaBlend = new BlendState(Blend.One, Blend.InverseSourceAlpha);
        NonPremultiplied = new BlendState(Blend.SourceAlpha, Blend.InverseSourceAlpha);
        Opaque = new BlendState(Blend.One, Blend.Zero);*/
        public static readonly BlendState Additive = new BlendState(Blend.SourceAlpha, Blend.One);
        public static readonly BlendState AlphaBlend = new BlendState(Blend.One, Blend.InverseSourceAlpha);
        public static readonly BlendState NonPremultiplied = new BlendState(Blend.SourceAlpha, Blend.InverseSourceAlpha);
        public static readonly BlendState Opaque = new BlendState(Blend.One, Blend.Zero);
    }
    public class RenderPipeline : IDisposable
    {
        public IntPtr handle;
        public Shader shader;
        public BlendState blendState;
        public bool depthWrite;
        public bool depthTest;
        public CullMode cullMode;
        public PrimitiveType primitiveType;
        public VertexDeclaration[] vertexDeclarations;

        public unsafe RenderPipeline(Shader shader,
            CullMode pipelineCullMode,
            PrimitiveType pipelinePrimitiveType,
            BlendState pipelineBlendState,
            bool testDepth,
            bool writeToDepth,
            VertexDeclaration[] vertexDeclarations)
        {
            Span<IntPtr> vertexHandles = stackalloc IntPtr[8];
            for (int i = 0; i < vertexDeclarations.Length; i++)
            {
                vertexHandles[i] = vertexDeclarations[i].handle;
            }
            fixed (IntPtr* ptr = vertexHandles)
            {
                handle = AstralCanvas.RenderPipeline_Init(shader.handle, pipelineCullMode, pipelinePrimitiveType, pipelineBlendState, testDepth, writeToDepth, ptr, (UIntPtr)vertexDeclarations.Length);
            }
            this.cullMode = pipelineCullMode;
            this.primitiveType = pipelinePrimitiveType;
            this.blendState = pipelineBlendState;
            this.depthTest = testDepth;
            this.depthWrite = writeToDepth;
            this.vertexDeclarations = vertexDeclarations;
        }
        public void Dispose()
        {
            if (handle != IntPtr.Zero)
            {
                AstralCanvas.RenderPipeline_Deinit(handle);
            }
            handle = IntPtr.Zero;
        }
    }
}

File: c-interface/src/RenderPipeline.cpp
#include "Astral.Canvas/Graphics/RenderPipeline.h"
#include "Graphics/RenderPipeline.hpp"

exportC void *AstralCanvasRenderPipeline_GetLayout(AstralCanvasRenderPipeline ptr)
{
    return ((AstralCanvas::RenderPipeline *)ptr)->layout;
}
exportC AstralCanvasShader AstralCanvasRenderPipeline_GetShader(AstralCanvasRenderPipeline ptr)
{
    return (AstralCanvasShader)&((AstralCanvas::RenderPipeline *)ptr)->shader;
}
exportC AstralCanvas_CullMode AstralCanvasRenderPipeline_GetCullMode(AstralCanvasRenderPipeline ptr)
{
    return (AstralCanvas_CullMode)((AstralCanvas::RenderPipeline *)ptr)->cullMode;
}
exportC AstralCanvas_PrimitiveType AstralCanvasRenderPipeline_GetPrimitiveType(AstralCanvasRenderPipeline ptr)
{
    return (AstralCanvas_PrimitiveType)((AstralCanvas::RenderPipeline *)ptr)->primitiveType;
}
exportC AstralCanvasBlendState AstralCanvasRenderPipeline_GetBlendState(AstralCanvasRenderPipeline ptr)
{
    AstralCanvasBlendState blendState = {
        (AstralCanvas_Blend)((AstralCanvas::RenderPipeline *)ptr)->blendState.sourceColorBlend,
        (AstralCanvas_Blend)((AstralCanvas::RenderPipeline *)ptr)->blendState.destinationColorBlend,
        (AstralCanvas_Blend)((AstralCanvas::RenderPipeline *)ptr)->blendState.sourceAlphaBlend,
        (AstralCanvas_Blend)((AstralCanvas::RenderPipeline *)ptr)->blendState.destinationAlphaBlend
    };
    return blendState;
}
exportC bool AstralCanvasRenderPipeline_IsDepthWrite(AstralCanvasRenderPipeline ptr)
{
    return ((AstralCanvas::RenderPipeline *)ptr)->depthWrite;
}
exportC bool AstralCanvasRenderPipeline_IsDepthTest(AstralCanvasRenderPipeline ptr)
{
    return ((AstralCanvas::RenderPipeline *)ptr)->depthTest;
}
exportC void AstralCanvasRenderPipeline_Deinit(AstralCanvasRenderPipeline ptr)
{
    ((AstralCanvas::RenderPipeline *)ptr)->deinit();
}

exportC AstralCanvasRenderPipeline AstralCanvasRenderPipeline_Init(
    AstralCanvasShader pipelineShader, 
    AstralCanvas_CullMode pipelineCullMode, 
    AstralCanvas_PrimitiveType pipelinePrimitiveType, 
    AstralCanvasBlendState pipelineBlendState, 
    bool testDepth, 
    bool writeToDepth, 
    AstralCanvasVertexDeclaration* vertexDeclarations,
    usize vertexDeclarationCount)
{
    AstralCanvas::RenderPipeline* result = (AstralCanvas::RenderPipeline*)GetCAllocator().Allocate(sizeof(AstralCanvas::RenderPipeline));
    AstralCanvas::BlendState blendState;
    blendState.sourceAlphaBlend = (AstralCanvas::Blend)pipelineBlendState.srcAlpha;
    blendState.destinationAlphaBlend = (AstralCanvas::Blend)pipelineBlendState.destAlpha;
    blendState.sourceColorBlend = (AstralCanvas::Blend)pipelineBlendState.srcColor;
    blendState.destinationColorBlend = (AstralCanvas::Blend)pipelineBlendState.destColor;

    collections::Array<AstralCanvas::VertexDeclaration *> vertexDecls = collections::Array<AstralCanvas::VertexDeclaration *>(GetCAllocator(), vertexDeclarationCount);
    for (usize i = 0; i < vertexDeclarationCount; i++)
    {
        vertexDecls.data[i] = (AstralCanvas::VertexDeclaration*)vertexDeclarations[i];
    }

    *result = AstralCanvas::RenderPipeline(GetCAllocator(), (AstralCanvas::Shader *)pipelineShader, (AstralCanvas::CullMode)pipelineCullMode, (AstralCanvas::PrimitiveType)pipelinePrimitiveType, blendState, testDepth, writeToDepth, vertexDecls);
    return result;
}